### ISSUE TITLE
Add a hybrid link-following option

### ISSUE THREAD

**Issue by berezovskyi** [eclipse/lyo.client#59]
This is a proposal for a new mode of processing Client responses (and, potentially, incoming Server payloads).

Right now, URIs in the object triple position will be unmarshalled into Link instances and must be followed manually. The user might not know if there is already a resource inlined in the body or whether it shall be fetched over HTTP.

I suggest to develop a new API to allow navigating those links seamlessly. This will be handy for inlining-intensive use-cases (including https://github.com/oasis-tcs/oslc-core/issues/17).

For now, I only have nav and navTry helper methods. They help looking up resources in the same Model. This is not enough, we need more API changes:

1. JMH and Client need to work with Jena Models.
1. Client responses should be some OSLC objects. The simplest way is to refactor JMH from a static class into a stateful object with a private Model field.

I am not sure how to add a capability of link-following and not break the abstraction (not to depend on OslcClient directly). I could also envision the introduction of typed Links.

Additionally, I have developed the following Kotlin classes to allow certain resources to be accompanied by other resources that shall be inlined with the "main" resource (not relying on the LocalResources in any way):

```kotlin
interface ResourceArray {
    fun asResourceArray(): Array<Object>
}


data class InstanceWithResources<T : IExtendedResource>(val instance: T,
                                                        val resources: Collection<IExtendedResource>) :
    ResourceArray {
    override fun asResourceArray(): Array<Object> {
        val resources: MutableList<Object> = ArrayList()
        resources.add(instance as Object)
        resources.addAll(resources)
        return resources.toTypedArray()
    }
}

data class InstanceMultiWithResources<T : IExtendedResource>(val instance: Collection<T>,
                                                             val resources: Collection<IExtendedResource>) :
    ResourceArray {
    override fun asResourceArray(): Array<Object> {
        val resources: MutableList<Object> = ArrayList()
        resources.addAll(instance as Collection<Object>)
        resources.addAll(resources)
        return resources.toTypedArray()
    }
}
```

**Comment by jamsden** [comment]
Let's use an example. Say I have a Requirement implementedBy ChangeRequest. I would expect requirement.getImplementedBy() to return a Set<ChangeRequest> if the change requests are actually read, and Set<Link> if not. Servers or particular OSLC queries that access the Requirement might or might not inline the ChangeRequests. If they do, then it would be useful to unmarshall them. If not, then I would expect the client to have to follow the link. 

A way to normalize this might be to have the requirement.getImplementedBy() always return Set<LInk> but the Link could be materialized with the actual object (in addition to its URL) or not depending on what was available when the RDF was parsed.

**Comment by berezovskyi** [comment]
Good example. Depending on how you define your POJOs, you will either have

    class Requirement {
       Set<ChangeRequest> getImplementedBy();
    }

or 

    class Requirement {
       Set<Link> getImplementedBy();
    }

If you look at https://github.com/eclipse/lyo.domains/blob/master/oslc-domains/src/main/java/org/eclipse/lyo/oslc/domains/rm/IRequirement.java#L256, it's the latter. What I effectively want is the easiest way to loop over the `ChangeRequest`s in question.

I currently have an idea of: (a) parametrising the Link and (b) implementing a Visitor pattern on it. Example:

    class Requirement {
       Set<TypedLink<ChangeRequest>> getImplementedBy();
    }

    class TypedLink<T> {
      T visit(OslcClient client);
    }

But then, this does not allow looking inside the Model of the response beforehand.

**Comment by berezovskyi** [comment]
The simplest extension to this is a common interface for the OslcClient and OslcResponse (yet to be created, that's the stateful JMH instance I talked about). Then this interface can use Composite pattern (or just a simple "fallback" Strategy pattern). This allows even to insert SPARQL lookup before an HTTP request (e.g. in case of caching).

```
class LinkVisitor {
  <T> T visit(TypedLink<T>);
}

OslcClient implements LinkVisitor

OslcResponse implements LinkVisitor

lyo.Store implements LinkVisitor

FallbackVisitor implements LinkVisitor {
FallbackVisitor(LinkVisitor primary, LinkVisitor fallback);
}
```

Does it make sense?

**Comment by berezovskyi** [comment]
We can also create standalone visitor classes without modifying the existing type hierarchy

**Comment by jamsden** [comment]
That could work. I was thinking of something more on the lines of a Link being able to lazily cache the thing it links to.

**Comment by berezovskyi** [comment]
Ok, let me see if I get this right. You were thinking of having something like this (WeakReference to prevent leaks when a Link will hold a reference forever, might need some extra thinking):

```
class TypedLink<T> {
private WeakReference<T> resource;

T getResource() {...}
```

This is handy from the usability perspective (just a simple getter call) but actually going to make our life harder. You say yourself that the client needs to decide how to fetch links. In my approach, the client supplies their LinkVisitor, for example by calling OslcResponse.createLinkVisitorWithFallback(new OslcClientLinkVisitor(client)). In your case (though I think all of these ideas can be combined in some way), the Lyo Client would actually have to call some setter first?

**Comment by jamsden** [comment]
I was thinking that if the original resource had the linked to resources in-lined, then the link could be hydrated the same time the containing resource is de-serialized. If it doesn't then the link only contains the URL. If the client app subsequently accessed an un-hydrated link, it could be lazily hydrated by doing a GET on the linked resource.

**Comment by berezovskyi** [comment]
Right, but then this would require us to embed this logic in the generated getters inside POJOs. This is directly the opposite of giving clients control over this process.

**Comment by berezovskyi** [comment]
But maybe we could implement something similar to this. Because such an interface would be really convenient to use.
